src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,389 @@
+        {
+            Unknown = 0,
+            Desktop = 1,
+            NetCore = 2,
+            SingleFile = 3,
+            NativeAOT = 4,
+        }
+
+        /// <summary>
+        /// This block is passed to the Watson DOTNET CLRMA provider and must not have any GC references
+        /// </summary>
+        private struct TriageBlockBuffer
+        {
+            public const int BufferSize = 8192;
+            public fixed byte Buffer[BufferSize];
+        }
+
+        private bool _comma;
+        private int _currentBufferIndex;
+        private int _reservedBuffer;
+        private static TriageBlockBuffer _triageBuffer;
+
+        public CrashInfo()
+        {
+            _comma = false;
+            _currentBufferIndex = 0;
+            _reservedBuffer = 1;            // Reserve 1 byte for closing }
+        }
+
+        /// <summary>
+        /// Writes the opening bracket and header to triage buffer
+        /// </summary>
+        /// <param name="reason"></param>
+        /// <param name="message"></param>
+        public void Open(RhFailFastReason reason, string message)
+        {
+            // Write the opening bracket and basic header which should never fail
+            bool success = WriteChar('{');
+            Debug.Assert(success);
+            success = WriteHeader(reason, message);
+            Debug.Assert(success);
+        }
+
+        /// <summary>
+        /// Write the closing bracket. The triage buffer is ready.
+        /// </summary>
+        public void Close()
+        {
+            // Write the closing bracket which should never fail since it was reserved
+            _reservedBuffer = 0;
+            bool success = WriteChar('}');
+            Debug.Assert(success);
+        }
+
+        /// <summary>
+        /// Write the exception information with fallbacks if the info doesn't fit in the fixed size triage buffer.
+        /// </summary>
+        /// <param name="exception">exception object</param>
+        public void WriteExceptionInfo(Exception exception)
+        {
+            if (!WriteBlock("exception", '{', '}', () => WriteException(exception, int.MaxValue, 500, int.MaxValue)))
+            {
+                // If the buffer isn't big enough to fit 500 stack frames, try limiting to 10
+                if (!WriteBlock("exception", '{', '}', () => WriteException(exception, int.MaxValue, 10, int.MaxValue)))
+                {
+                    // If that fails, try limiting the size of the stack frame method names to 100 bytes
+                    WriteBlock("exception", '{', '}', () => WriteException(exception, int.MaxValue, 10, 100));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the rendered crash triage buffer address
+        /// </summary>
+        public IntPtr GetTriageBuffer(out int size)
+        {
+            size = _currentBufferIndex;
+            return new IntPtr(Unsafe.AsPointer<TriageBlockBuffer>(ref _triageBuffer));
```

Comment:
Statics in .NET are not guaranteed to be pinned. This code has GC hole.

(I think that this specific static is going to be pinned due to native AOT implementation details, but it is a fragile assumption to make.)

Reply:
What is the best way to pin/get the address in Native AOT?  I assume using a fixed statement will only pin it while in the block.

Reply:
There is nothing special in native AOT to do this. It has same tools for pinning as CoreCLR - GCHandles, pinned object heap, etc.

This comment is no longer relevant given https://github.com/dotnet/runtime/pull/88273/files#r1249902094